src/integration/resources/inprocess/coordinator.go

```diff
@@ -0,0 +1,278 @@
+	"github.com/m3db/m3/src/query/generated/proto/prompb"
+	"github.com/m3db/m3/src/query/server"
+	xconfig "github.com/m3db/m3/src/x/config"
+)
+
+type coordinator struct {
+	cfg     config.Configuration
+	client  common.CoordinatorClient
+	logger  *zap.Logger
+	tmpDirs []string
+
+	interruptCh chan<- error
+	shutdownCh  <-chan struct{}
+}
+
+// CoordinatorOptions are options for starting a coordinator server.
+type CoordinatorOptions struct {
+	// Logger is the logger to use for the coordinator. If not provided,
+	// a default one will be created.
+	Logger *zap.Logger
+}
+
+// NewCoordinatorFromConfigFile creates a new in-process coordinator based on the config file
+// and options provided.
+func NewCoordinatorFromConfigFile(pathToCfg string, opts CoordinatorOptions) (resources.Coordinator, error) {
+	var cfg config.Configuration
+	if err := xconfig.LoadFile(&cfg, pathToCfg, xconfig.Options{}); err != nil {
+		return nil, err
+	}
+
+	return NewCoordinator(cfg, opts)
+}
+
+// NewCoordinatorFromYAML creates a new in-process coordinator based on the YAML configuration string
+// and options provided.
+func NewCoordinatorFromYAML(yamlCfg string, opts CoordinatorOptions) (resources.Coordinator, error) {
+	var cfg config.Configuration
+	if err := yaml.Unmarshal([]byte(yamlCfg), &cfg); err != nil {
+		return nil, err
+	}
+
+	return NewCoordinator(cfg, opts)
+}
+
+// NewCoordinator creates a new in-process coordinator based on the configuration
+// and options provided.
+func NewCoordinator(cfg config.Configuration, opts CoordinatorOptions) (resources.Coordinator, error) {
+	// Replace any "0" ports with an open port
+	cfg, err := updateCoordinatorPorts(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	// Replace any "*" filepath with a temporary directory
+	cfg, tmpDirs, err := updateCoordinatorFilepaths(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	// Configure logger
+	if opts.Logger == nil {
+		opts.Logger, err = zap.NewDevelopment()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Get HTTP port
+	_, p, err := net.SplitHostPort(cfg.ListenAddressOrDefault())
+	if err != nil {
+		return nil, err
+	}
+
+	port, err := strconv.Atoi(p)
+	if err != nil {
+		return nil, err
+	}
+
+	// Start the coordinator
+	coord := &coordinator{
+		cfg: cfg,
+		client: common.NewCoordinatorClient(common.CoordinatorClientOptions{
+			Client:    &http.Client{},
+			HTTPPort:  port,
+			Logger:    opts.Logger,
+			RetryFunc: retry,
+		}),
+		logger:  opts.Logger,
+		tmpDirs: tmpDirs,
+	}
+	coord.start()
+
+	return coord, nil
+}
+
+func (c *coordinator) start() {
+	interruptCh := make(chan error, 1)
+	shutdownCh := make(chan struct{}, 1)
+
+	go func() {
+		server.Run(server.RunOptions{
+			Config:      c.cfg,
+			InterruptCh: interruptCh,
+			ShutdownCh:  shutdownCh,
+		})
+	}()
+
+	c.interruptCh = interruptCh
+	c.shutdownCh = shutdownCh
+}
+
+func (c *coordinator) GetNamespace() (admin.NamespaceGetResponse, error) {
```

Comment:
Are these implementing a specific interface? If so can you add the build check at the bottom of the file:

var (
  _ SomeInterface =  &coordinator{}
)

Reply:
FWIW, you get this for free with the constructors:

```go
func NewCoordinator(cfg config.Configuration, opts CoordinatorOptions) (resources.Coordinator, error) {
	// ... snip

	coord := &coordinator{
		cfg: cfg,
		client: common.NewCoordinatorClient(common.CoordinatorClientOptions{
			Client:    &http.Client{},
			HTTPPort:  port,
			Logger:    opts.Logger,
			RetryFunc: retry,
		}),
		logger:  opts.Logger,
		tmpDirs: tmpDirs,
	}
	coord.start()

        // This line fails the build
	return coord, nil
}
```

Is the suggestion so it's easier to recognize the interface being implemented?